### ISSUE TITLE
Improved loading indicator

### ISSUE THREAD

**Issue by tuanchoang** [jupyterlite/jupyterlite#619]
### Problem

Right now, when JupyterLite is loading/launching in the browser, it could take a while to load and the unsuspecting user could be waiting somewhere ~10 seconds wondering if anything is happening before loading completes.

### Proposed Solution

JupyterLite could provide a better user experience by visualizing a progress bar and/or loading indicator while it is launching. This gives the user a sense of time and provides context that JupyterLite is loading and the expectation is to wait for it to finish loading.

**Comment by jtpio** [comment]
Thanks @tuanchoang for the suggestion :+1: 

> Right now, when JupyterLite is loading/launching in the browser, it could take a while to load and the unsuspecting user could be waiting somewhere ~10 seconds wondering if anything is happening before loading completes.

Just to clarify, does that happen during the initial page load? Or are you referring to the slow loading times of the Python kernel when a notebook is opened?

**Comment by tuanchoang** [comment]
@jtpio Sorry I could have been clearer. :)

Yeah, I'm referring to the initial page load. To the end-user, it could look like nothing is happening when in actuality JupyterLite hasn't finished loading.

**Comment by jtpio** [comment]
Alright thanks!

Maybe there could then be some kind of static spinner / loaded that could be loaded on the page on startup: https://github.com/jupyterlite/jupyterlite/blob/main/app/lab/index.template.html

And removed when the app is ready.

**Comment by tuanchoang** [comment]
Exactly, that'd be perfect! Just some sort of indicator to let the user know to wait and/or that it's loading.

**Comment by bollwyvl** [comment]
Sounds great: lots of other stuff going on, so I'm not sure who would do this, and therefore, when this would land. Some thoughts:

This is one of the few things that _can't_ be an extension, and probably _does_ need to be replaceable, if not particularly _configurable_. My gut feel: 
- make a new file in the root called e.g. `loading.html`
  - copy it from the `lite_dir` or use a CLI option `--loading-html`
- load that as an iframe in the DOM of each `index.html`, no JS, no stacking
- in `config-utils.js`
  - start a `message` listener which can clear the iframe
    - also have a timeout, so if something _does_ go wrong
      - this _can_ be configurable
- in each of our base applications, `window.postMessage` right _before_ the moon animation starts
  - themes should still control those

`postMessage` is going to be A Thing, anyway (I've already hacked a couple of them up) so maybe we need to think about that some.

**Comment by jtpio** [comment]
> load that as an iframe in the DOM of each `index.html`, no JS, no stacking

Maybe this could only be for the lab app? Not sure a loading indicator would feel natural for other apps like retro and repl.

**Comment by bollwyvl** [comment]
>  Not sure a loading indicator would feel natural for other apps

ha: i think it would be _extra_ important in these situations. Bottom line: a blank screen looks broken. i think the goal is  **pixels on page ASAP** so people don't bounce.

Indeed, in addition to these, we might want to consider a "lobby" that would defer the loading of anything until actually clicked, which would help with focus. but this is for another day...

> what kind of indicator to use,

gah, the blur! my eyes! but seriously, i did expect the bulb to focus in, or something.

semi-related: doing an actual, accurate "55% loading/.." indicator is probably not worth the effort... i'm thinking the default is something looping, unobtrusive, portable, but most of all _small_ (and therefore fast): I recently used something along the lines of [these](https://projects.lukehaas.me/css-loaders/)... just a little thing, centered on the page.

Point is: as long as it's super small, and easy to change by replacing one file, everybody wins.

**Comment by certik** [comment]
For the SymPy live (https://www.sympy.org/en/shell.html) experience, I would like to see a clear indicator when the Python/JupyterLite itself is loading (~5s), and then another one when sympy itself is loading (~5s).

**Comment by asmeurer** [comment]
Even when a calculation is happening it's not obvious. There's just a `[*]` by the prompt. Although maybe this is more of a general Jupyter problem.

**Comment by kolibril13** [comment]
In https://jupyterlite-sphinx.readthedocs.io/en/latest/directives/retrolite.html#
there is a nice logo that is zooming in and out on a 1-second period when the initial page load starts. 
As jupyterlite-sphinx and jupyterlite are in the same ecosystem, I think it would be great to have this same logo in Jupyterlite as well:
<img width="361" alt="image" src="https://github.com/jupyterlite/jupyterlite/assets/44469195/d851dce5-d366-43b2-aa6d-2f5b401dfb40">

**Comment by jtpio** [comment]
This upstream issue in JupyterLab may also be relevant: https://github.com/jupyterlab/jupyterlab/issues/5463

**Comment by psychemedia** [comment]
I'm hitting this as an issue when loading a jupyterlite site served from an html5.zip interactive package in a Moodle VLE environment; the load time can be 10s of seconds on first run (once everything is cached in the browser it's not so bad) and the blank screen leaves users uncertain as to whether the page is actually loading or not...

**Comment by jtpio** [comment]
https://github.com/jupyterlite/jupyterlite/pull/1636 adds a loading indicator for the JupyterLab app:

https://github.com/user-attachments/assets/a46f7434-b4d0-4a68-b71c-8227f1a74f4c

If you would like to try it: https://jupyterlite--1636.org.readthedocs.build/en/1636/_static/lab/index.html
